components/logins/android/src/test/java/mozilla/appservices/logins/DatabaseLoginsStorageTest.kt

```diff
@@ -91,4 +98,86 @@ class DatabaseLoginsStorageTest : LoginsStorageTest() {
 
         finishAndClose(test)
     }
+
+    @Test
+    fun testMetricsGathering() {
```

Comment:
*I AM CRYING*.

This is **great**, you're writing unit-tests for telemetry. The Glean team really appreciates this :)

Reply:
Heh, well then, you'll be pleased to hear that these tests actually helped uncover a serious bug in the metrics gathering logic :-)

Previously the `errCount` metrics were not being gathered correctly, because I was trying to catch them inside the `rustCall` wrapper which is responsible for turning error codes into exceptions. We could easily have shipped that to production if there wasn't a really obvious way to test that the metrics gathering was actually working.

Some quick feedback on the process of adding these tests, while it's fresh in my head:

* I foolishly tried to add them by just figuring it out based on the available `test*` methods that I noticed in the glean class definitions. I got a lot of inscrutable errors when running `./gradlew logins:test`, but once I figured out to run `./gradlew logins:test --info` then glean provided helpful error messages to get me up and running. The testing section in the glean book was also helpful, once I actually stopped to read it.
* I originally left out the `GleanTestRule` because of the above. Glean gave me an error about needing to call `Glean.shared.enableTestingMode()`, which was helpful. But the latest a-c tells me that this is deprecated in favour of `GleanTestRule`. It may be helpful to update glean's error message to reflect this.
* I had to upgrade to robolectric 4.0 in order to get support for `elapsedRealtimeNanos` in the tests, which took me a bit of googling to figure out.
* It may or may not make sense based on your mental model of things, but I found myself wanting a `LabeledMetric::testHasValue()` helper function that will test whether a value has been recorded for any label.

Overall though: thanks for making it easy 👍

Reply:
That's a great motivational comment, thank you very much! We filed bugs for all this problems: [wrong error message](https://bugzilla.mozilla.org/show_bug.cgi?id=1600005), [robolectric requirement not being advertised](https://bugzilla.mozilla.org/show_bug.cgi?id=1600006) and the [request for the new testing API](https://bugzilla.mozilla.org/show_bug.cgi?id=1600008).

Reply:
I love this comment thread! ❤️😭